### PR TITLE
fix/143: Cancelling Google Pay on checkout shows validation errors

### LINKED ISSUE #143: [#133 followup] Cancelling Google Pay on checkout shows validation errors.
### Describe the bug

Cancelling Google Pay payments shows validation errors on the checkout form prior to the purchaser pressing the place order button. While these validation errors are an unpleasant experience, they do not prevent the purchaser from checking out. 

When cancelling Google Pay payment forms, the validation errors should not display on the form until the user presses the submit button.

### To reproduce

1. Testing the front end should take place in Chrome while logged in to a Google account with Google Pay configured
2. Configure the extension to enable digital wallets
3. Create a simple product
4. Ensure the checkout page is using blocks
5. In Chrome, visit the product page and add to cart
6. Go to the checkout (blocks) page
7. Click the Google Pay button
8. Once Google Pay has loaded, click the `x` icon to cancel the transaction
9. The checkout page will show validation errors for any required fields that remain empty

**Screenshots**

Checkout form after cancelling Google Pay

334673725-dc0e4341-061a-4aa1-a5f3-b44ac40b41d7

### Expected behavior

The checkout form should only show validation errors after the user has interacted with the fields or clicked the submit button.

**Environment (please complete the following information):**
 - WordPress 6.5.3
 - WooCommerce 8.9.1
 - WooCommerce Square: Branch `fix/133-checkout-blocked` @ 8cab6393f508307bb016dfeb1c9e5b6b54389ff0
 - Browser Chrome 125
 - Any other plugins installed

### PR DESCRIPTION
**Pull request by Sidsector9**
### All Submissions:

* [ ] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:

Closes #143 

This PR does 2 things:

- Fixes the issue #143
- Refactors the implementation of Express Payment to align with other Woo gateways.

### Steps to test the changes in this Pull Request:

Test Digital Wallets features.

### Changelog entry

> Fix - Cancelling Google Pay on checkout shows validation errors.

### CODE REVIEW

File: assets/blocks/digital-wallets/content.js [hunk 1]
```diff
@@ -6,48 +6,104 @@ import { useEffect, useState } from '@wordpress/element';
...
+	);
+
+	const applePayExpressButton = !isApplePayDisabled && (
+		<div // eslint-disable-line jsx-a11y/click-events-have-key-events
+			tabIndex={0}
+			role="button"
+			ref={applePayRef}
+			onClick={() => setClickedButton(applePay)}
+		></div>
+	);
```
Comment (peterwilsoncc) [inline assets/blocks/digital-wallets/content.js]: This ensures the Apple Pay button displays correctly and resolves a JS error.

You'll need to add something to ensure the button is hidden on render.


```suggestion
	const applePayExpressButton = !isApplePayDisabled && (
		<div // eslint-disable-line jsx-a11y/click-events-have-key-events
			tabIndex={0}
			role="button"
			ref={applePayRef}
			onClick={() => setClickedButton(applePay)}
			className="apple-pay-button wc-square-wallet-buttons"
		>
			<span className="text"></span>
			<span className="logo"></span>
		</div>
	);
```

File: assets/blocks/digital-wallets/content.js [hunk 2]
```diff
@@ -6,48 +6,104 @@ import { useEffect, useState } from '@wordpress/element';
...
-			setPayments(__payments);
-		} catch (e) {
-			console.error(e);
-		}
-	}, []);
+		setExpressPaymentError('');
+		onClick();
+
+		(async () => {
+			const __tokenResult = await tokenize(clickedButton);
```
Comment (peterwilsoncc) [inline assets/blocks/digital-wallets/content.js]: When clicking the Apple Pay button I am seeing the error `Unhandled Promise Rejection: UnexpectedError: An unexpected error occurred while using Apple Pay.` with a reference to this line.

**Review by peterwilsoncc — COMMENTED** [review]
I've added some notes inline. 

I am getting a JavaScript error for Apple Pay in Safari so the payment form fails to open but have yet to figure out the cause. 

For Google Pay in Safari it re-introduces a bug that was fixed earlier in which the entire page is redirected to the Google Pay form. I tried to find the commit in which it was fixed but without success


https://github.com/user-attachments/assets/709a919b-79da-4795-a9cc-ffa7f1640663

**Comment by Sidsector9** [comment]
@peterwilsoncc I've fixed the issues you pointed, requesting a re-review.

File: assets/blocks/digital-wallets/hooks.js [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+			applePayRef.current.classList.add(
+				'wc-square-wallet-button-with-text'
+			);
+		}
+
+		applePayRef.current.style.cssText += `-apple-pay-button-type: ${type};`;
+		applePayRef.current.style.cssText += `-apple-pay-button-style: ${color};`;
+		applePayRef.current.style.display = 'block';
+		applePayRef.current.classList.add(`wc-square-wallet-button-${color}`);
+	}, [applePayRef]);
```
Comment (peterwilsoncc) [inline assets/blocks/digital-wallets/hooks.js]:
```suggestion
	}, [applePayRef, applePay]);
```

File: assets/blocks/digital-wallets/hooks.js [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+				await payments.applePay(paymentRequest);
+				const __applePay = await payments.applePay(paymentRequest);
+
+				setApplePay(__applePay);
+			} catch (e) {}
+		})();
+	}, [payments, paymentRequest]);
+
+	useEffect(() => {
+		if (!applePayRef?.current) {
```
Comment (peterwilsoncc) [inline assets/blocks/digital-wallets/hooks.js]: Required with above suggestion to ensure the Apple Pay button is only shown in Safari.

```suggestion
		if (!applePayRef?.current || !applePay) {
```

**Review by peterwilsoncc — COMMENTED** [review]
This is working great, I've made a suggestion inline to prevent the Apple button displaying in non-Safari Browsers.

The one bug I am seeing with Apple Pay is when there are no cards defined. If I cancel the payment attempt, the Apple and Google Pay buttons remain locked. 

I suspect a different event is fired for cancelled payment requests as opposed to invalid requests. I think this might be an existing bug so if it is, I'd recommend opening a follow up ticket rather than letting it block you on this one

locked

The validation error fixes are looking great and working as expected. Nice work.

File: assets/blocks/digital-wallets/hooks.js (hunk 4 above)
Comment (Sidsector9) [reply assets/blocks/digital-wallets/hooks.js]: @peterwilsoncc I've added these suggestions.

**Review by peterwilsoncc — APPROVED** [review]
Thanks Sid, this looks good and is working as expected.

**Testing notes**

In Safari, checkout page (blocks):
* Click Apple Pay button
* Cancel
* Validation errors are not shown (as expected)
* Click Google Pay button
* Cancel
* Validation errors are not shown (as expected)
* Click Apple Pay button
* Proceed through payment
* Order processed as expected
* Re-order, return to checkout
* Click Google Pay button 
* Proceed through payment
* Order processed as expected

In Safari, cart page (blocks)
* Click Apple Pay button
* Cancel
* Click Google Pay button
* Cancel
* Click Apple Pay button
* Proceed through payment
* Order processed as expected
* Re-order, return to cart
* Click Google Pay button 
* Proceed through payment
* Order processed as expected

In Chrome, checkout page (blocks):
* Apple Pay button not shown
* Click Google Pay button
* Cancel
* Validation errors are not shown (as expected)
* Click Google Pay button 
* Proceed through payment
* Order processed as expected

In Chrome, cart page (blocks)
* Apple Pay button not shown
* Click Google Pay button
* Cancel
* Click Google Pay button 
* Proceed through payment
* Order processed as expected

**Comment by qasumitbagthariya** [comment]
### QA Update ✅ 

-------------

I have verified this PR in the `fix/143` branch has been fixed and is functioning as intended.

I tested the following on this branch:
- Google Pay at checkout page ✅ 
- Validation errors are not shown ✅ 

**Testing Environment** 

<Details>

- WordPress: 6.6.1
- Theme: Storefront 4.6.0
- Theme: Twenty Twenty-Four 1.2
- WooCommerce - 9.1.4
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Branch: fix/143

</Details>

https://github.com/user-attachments/assets/0bfb9972-2be9-45fa-862f-4b8c80104300


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [ ] Yes
- [x] Not Required/Applicable for this PR
------------------------------------------------

File: src/blocks/digital-wallets/hooks.js [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+	useEffect(() => {
+		if (!(payments && paymentRequest)) {
+			return;
+		}
+
+		(async () => {
+			try {
+				await payments.googlePay(paymentRequest);
+				const __googlePay = await payments.googlePay(paymentRequest);
+				await __googlePay.attach(googlePayRef.current, {
```
Comment (Mayisha) [inline src/blocks/digital-wallets/hooks.js]: `await payments.googlePay(paymentRequest);` is twice here.

File: src/blocks/digital-wallets/hooks.js [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+	useEffect(() => {
+		if (!(payments && paymentRequest)) {
+			return;
+		}
+
+		(async () => {
+			try {
+				await payments.applePay(paymentRequest);
+				const __applePay = await payments.applePay(paymentRequest);
+
```
Comment (Mayisha) [inline src/blocks/digital-wallets/hooks.js:160]: `await payments.applePay(paymentRequest)` is also called twice similar to `googlePay`. Is this intentional?

File: src/blocks/digital-wallets/hooks.js [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+		() =>
+			onPaymentSetup(() => {
+				if (!tokenResult) {
+					return;
+				}
+
+				async function handlePaymentProcessing() {
+					let response = { type: emitResponse.responseTypes.SUCCESS };
+
+					if (!tokenResult) {
```
Comment (Mayisha) [inline src/blocks/digital-wallets/hooks.js]: Is this check going to work as we are already returning from line `212` above for the same condition?

**Comment by vikrampm1** [comment]
@Mayisha this PR is ready for your review!

**Review by Mayisha — APPROVED** [review]
Changes look good to me and it fixed the original issue.

- I can reproduce the issue in `trunk`. When I cancel Google Pay on the block checkout page, it triggers the form validation.

Screenshot 2024-08-07 at 4 24 08 PM

- In this branch canceling Google Pay does not trigger the form validation on the block checkout page.

I just marked a few lines in my comment below which look like typos to me.

File: src/blocks/digital-wallets/hooks.js (hunk 7 above)
Comment (Sidsector9) [reply src/blocks/digital-wallets/hooks.js]: Thanks for catching these, I've made the changes in [33a8a84](https://github.com/woocommerce/woocommerce-square/pull/188/commits/33a8a843ac01c230a2ff4533c300f44278ddac58)

Comment (Mayisha) [reply src/blocks/digital-wallets/hooks.js]: Thanks @Sidsector9 for the update. The PR is already approved. 
I have tested this branchagain and confirmed that
- Canceling Google Pay does not trigger the form validation on the block checkout page.
- I can successfully purchase a product with Google Pay.

**Comment by qasumitbagthariya** [comment]
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.5.5, npm version 8.19.4, node version 16.20.0)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

**Testing Environment** 

<Details>

- WordPress: 6.6.1
- Theme: Storefront 4.6.0
- Theme: Twenty Twenty-Four 1.2
- WooCommerce - 9.1.4
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Branch: smoke-testing

</Details>



Next Step- Ready to Merge 🚀

**Pull request merged**